build(docs-infra): upgrade @angular/cli to 8.0.0-beta.15, @angular/* to 8.0.0-beta.12, @angular/material to 8.0.0-beta.2

gkalpak opened the pull request:
This PR upgrades `@angular/{*, material, cli}` to current `@next` versions, enables differential loading, enables some more strict TypeScript checks for angular.io and aligns the config files closer to what the cli would generate today.
See individual commits for more details.

Blocked on #29976.

filipesilva reviewed (commented):
Can you also re-enable these tests please? https://github.com/angular/angular/pull/28685/commits/4d37a060e1b99dc58cb2d2c0c1b61bef41ce9a4c

I know we have already talked about it but wanted to leave a comment here so I don't forget to check :D

gkalpak commented:
These tests are not related to the aio app's cli version (which is independent of the version used in docs examples).
I have added a TODO on my list to enable those tests in a subsequent PR, while upgrading the dependencies used in docs examples (including the cli).

mary-poppins commented:
You can preview bf28bcf at https://pr29926-bf28bcf.ngbuilds.io/.

mary-poppins commented:
You can preview e83ef61 at https://pr29926-e83ef61.ngbuilds.io/.

mary-poppins commented:
You can preview af27758 at https://pr29926-af27758.ngbuilds.io/.

Splaktar commented on aio/browserslist:
```diff
@@ -0,0 +1,13 @@
+Firefox ESR
+not dead
+IE 9-11 # For IE 9-11 support, remove 'not'.
+Chrome 41 # For Googlebot support, remove 'not'.
```
Consider removing the `, remove 'not'` from these two lines since it looks like the `not` was already removed?

Also does the `# For IE 9-11 support` comment add any value?

Consider adding a reference to Chrome 41 to the Googlebot comment above and removing the comment on this line?

gkalpak replied:
This file is generated by the cli, so I only wanted to make the absolutely necessary changes (i.e. remove the `not`).
I'll remove the `, remove not` part to avoid confusing.

It would be nice to incorporate the rest of the suggestions (which I totally agree with) in cli's [browserslist.template][1]. Can you create an issue or PR there?

[1]: https://github.com/angular/angular-cli/blob/b95933a57/packages/schematics/angular/application/files/browserslist.template

Splaktar replied:
I looked at the template and it seems to make sense there. It's mostly when you remove the `not` in the front that things get a little odd. I think that the latest update here looks good based on the desire to keep it close to vanilla CLI.

Splaktar commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -29,8 +29,10 @@ export class SearchBoxComponent implements OnInit {
 
-  @ViewChild('searchBox') searchBox: ElementRef;
+  @ViewChild('searchBox', { static: true }) searchBox: ElementRef;
+  // tslint:disable-next-line: no-output-on-prefix
```
Too much work/breakage to change this to `search` or `searchEntry`?

gkalpak replied:
Yeah, I wanted to avoid too much noise (in an already noisy PR).

Splaktar commented on aio/src/app/shared/scroll.service.spec.ts:
```diff
@@ -230,7 +230,7 @@ describe('ScrollService', () => {
   describe('#scrollToTop', () => {
     it('should scroll to top', () => {
-      const topOfPageElement = <Element><any> new MockElement();
+      const topOfPageElement = new MockElement() as any as Element;
```
It would be nice to have a comment here for why `MockElement` can't be directly cast to `Element`.

gkalpak replied:
Hm...how so? Is this unexpected (given `MockElement` is just a simple class, totally unrelated to `Element`)?

Splaktar commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -16,6 +16,7 @@ export class SelectComponent implements OnInit {
   @Input()
   options: Option[];
 
+  // tslint:disable-next-line: no-output-native
```
Too much work/breakage to change this to `optionChange`?

gkalpak replied:
Yeah, I wanted to avoid too much noise (in an already noisy PR).

Splaktar commented on aio/src/app/shared/toc.service.ts:
```diff
@@ -57,6 +57,7 @@ export class TocService {
     const div: HTMLDivElement = this.document.createElement('div');
     div.innerHTML = heading.innerHTML;
     const anchorLinks: NodeListOf<HTMLAnchorElement> = div.querySelectorAll('a');
+    // tslint:disable-next-line: prefer-for-of
```
Should we consider disabling this rule?

petebacondarwin replied:
Or just change to use `for of`?

gkalpak replied:
@petebacondarwin: TypeScript complains when iterating over `NodeList`s, unless you include the [dom.iterable][1] lib. Including the lib then requires compiling with `--downlevelIteration` for ES5 (which I assume would result in more verbose code).

@Splaktar: My rule of thumb was that if the rule only requires a couple of `disable` comments, then leave it. (To be clear, this rule is added by default for a new cli-geenrated app).

##
I just realized I can switch to for-of by wrapping the `NodeList`s in `Array.from()`. Everybody wins \o/ 😛 

[1]: https://github.com/Microsoft/TypeScript/blob/2eea21636/lib/lib.dom.iterable.d.ts

Splaktar commented on aio/tools/firebase-test-utils/FirebaseGlob.ts:
```diff
@@ -1,5 +1,10 @@
 import * as XRegExp from 'xregexp';
 
+// The `XRegExp` typings are not accurate and.
```
and?

gkalpak replied:
...and period 😛

Splaktar reviewed (approved):
This is full of awesome! `strict` ✅

I only had some questions and minor nits.

petebacondarwin commented on aio/src/app/search/search.worker.ts:
```diff
@@ -1,3 +1,4 @@
+/// <reference lib="webworker" />
```
Interesting...

gkalpak replied:
This is what was decided in angular/angular-cli#14188 (as the best way to support both IDEs and correct compilation (and avoid unnecessary config interdependencies, which are hard and error-prone to manage; especially in larger projects)).

mary-poppins commented:
You can preview 3122d82 at https://pr29926-3122d82.ngbuilds.io/.

gkalpak commented on aio/angular.json:
```diff
@@ -59,8 +59,7 @@
               "src/styles.scss"
             ],
-            "scripts": [],
-            "es5BrowserSupport": true
```
Removed in angular/angular-cli@b976a3e72.

mary-poppins commented:
You can preview 8b13fcf at https://pr29926-8b13fcf.ngbuilds.io/.

gkalpak commented:
@Splaktar, @petebacondarwin: Thx for the reviews. Addressed or responded to comments.

mary-poppins commented:
You can preview f18c685 at https://pr29926-f18c685.ngbuilds.io/.

mary-poppins commented:
You can preview 8c6fbc5 at https://pr29926-8c6fbc5.ngbuilds.io/.

mary-poppins commented:
You can preview 36e8af1 at https://pr29926-36e8af1.ngbuilds.io/.

mary-poppins commented:
You can preview 0c1938d at https://pr29926-0c1938d.ngbuilds.io/.

gkalpak committed:
build(docs-infra): remove obsolete `ie-polyfills.js`

Now that defferential loading it supported by the cli, remove the
obsolete `ie-polyfills.js` (and associated dependencies).

All polyfills in `ie-polyfills.js` are now included in the
[polyfills-es5][1] bundle, except for `classlist.js`, that is only
needed in order to support `NgClass` on SVG elements, which we don't
use.

[1]: https://github.com/angular/angular-cli/blob/b95933a57/packages/angular_devkit/build_angular/src/angular-cli-files/models/es5-polyfills.js

gkalpak committed:
build(docs-infra): upgrade tslint to 5.15.0 and codelyzer to 5.0.0

This commit also changes the `tslint.json` config file to (reasonably
closely) match what the cli would generate for a new app.

gkalpak committed:
build(docs-infra): make type-checking stricter by enabling `noImplicitAny`

gkalpak committed:
build(docs-infra): switch from no-op `no-jasmine-focus` tslint rule to `ban`

The `no-jasmine-focus` rule has been removed from
`vrsource-tslint-rules` [since version 5.8.0][1] (theoretically;
practically it remained [until version 5.8.2][2]).

This commit removes the non-existent rule (and the obsolete dependency)
and uses tslint's `ban` rule instead (as recommended).

[1]: https://github.com/vrsource/vrsource-tslint-rules/commit/477f622cb2b6adb66eecce69a122834920cee7a0#diff-04c6e90faac2675aa89e2176d2eec7d8R162
[2]: https://github.com/vrsource/vrsource-tslint-rules/pull/13#issuecomment-442053340

mary-poppins commented:
You can preview f43a5ea at https://pr29926-f43a5ea.ngbuilds.io/.

IgorMinar commented on aio/scripts/_payload-limits.json:
```diff
@@ -2,9 +2,12 @@
+        "runtime-es2015": 2986,
+        "main-es5": 501591,
+        "main-es2015": 440628,
+        "polyfills-es5": 128751,
```
we doubled the size of the polyfills compared to the previous version. do we know what's causing it? this seems like a size regression to fix.

// @filipesilva

filipesilva replied:
@alan-agius4 @clydin I think you both looked at the polyfills changes with diff. loading, can you shed some light here (and on the comment below)?

gkalpak replied:
The `polyfills-es5` file includes the ES2015 polyfills (i.e. former `polyfills.js`) plus the ES5 specific polyfills, so it is expected to be larger.
@IgorMinar, it is basically combining the former `polyfills.js` with our [ie-polyfills.js][1].

[1]: https://github.com/angular/angular/blob/a400429fa/aio/src/ie-polyfills.js

clydin replied:
Was just typing a similar response.  @gkalpak is correct.  The CLI now automatically handles (and removes the need for) the `ie-polyfills` file previously used in AIO.

IgorMinar replied:
Ah. That makes sense. It’s a bit surprising but with an explanation it makes sense

IgorMinar commented on aio/scripts/_payload-limits.json:
```diff
@@ -2,9 +2,12 @@
+        "main-es5": 501591,
+        "main-es2015": 440628,
+        "polyfills-es5": 128751,
+        "polyfills-es2015": 59557
```
there is no size reduction compared to v7 es5 polyfill bundle? that seems odd, zone.js alone should have helped to drop the size now that it should be differentially served. is that happening?

gkalpak replied:
@clydin, do you know if Zone.js' differentiality is taken advantage of here?

clydin replied:
It's currently not implemented in the CLI.

IgorMinar replied:
Did we forget to create a Jira item for this? We discussed it but might have forgotten to schedule. Can you please add to the backlog for v8? Thanks

clydin replied:
Pushed a PR to add it in: https://github.com/angular/angular-cli/pull/14273

IgorMinar commented on aio/scripts/_payload-limits.json:
```diff
@@ -2,9 +2,12 @@
-        "polyfills": 59197
+        "runtime-es5": 2980,
+        "runtime-es2015": 2986,
+        "main-es5": 501591,
```
we gained 20k :-( not great, but not the highest priority problem to attack, but I wonder if that means that we are leaking something that is also being leaked in the es2015 build. we might want to investigate this a bit after RC to at least understand where these 20k are coming from.

filipesilva replied:
@gkalpak can you make an issue to look into this for AIO? I can help there.

gkalpak replied:
@IgorMinar, we did not gain 20K here (501585 --> 501591).
The 20K has been gained while upgrading from @angular/*@7.2.9 to @angular/*@8.0.0-beta.13 (a8ee6e3c2).

I assume that is expected. LMK if we still need to investigate.

IgorMinar reviewed (approved):
the rest looks great. thank you for testing all of these bits together.

IgorMinar reviewed (commented):
since the package.json is already conflicting, could you please:
- rebase
- bump the version numbers to the latest betas for framework and cli?

thank you!

gkalpak committed:
build(docs-infra): upgrade @angular/* to 8.0.0-beta.14

gkalpak committed:
fixup! build(docs-infra): upgrade @angular/* to 8.0.0-beta.14

gkalpak committed:
build(docs-infra): upgrade @angular/material to 8.0.0-beta.2

gkalpak committed:
fixup! build(docs-infra): upgrade @angular/material to 8.0.0-beta.2

gkalpak committed:
fixup! build(docs-infra): upgrade @angular/cli to 8.0.0-beta.18

gkalpak committed:
build(docs-infra): upgrade @angular/cli to 8.0.0-beta.18

This commit also changes the config files and their layout to
(reasonably closely) match what the cli would generate for a new app.

Related Jira issue: [TOOL-815](https://angular-team.atlassian.net/browse/TOOL-815)

gkalpak commented:
@IgorMinar: Upgrade @angular/* to 8.0.0-beta.14, but left @angular/cli to 8.0.0-beta.15, because from 8.0.0-beta.16 onwards (till 8.0.0-beta.18), I get the following error during build:

```
Unknown browser kaios
BrowserslistError: Unknown browser kaios
    at checkName (...\angular\aio\node_modules\browserslist\index.js:171:20)
    at Function.select (...\angular\aio\node_modules\browserslist\index.js:881:18)
    at ...\angular\aio\node_modules\browserslist\index.js:215:33
    at Array.reduce (<anonymous>)
    at resolve (...\angular\aio\node_modules\browserslist\index.js:197:18)
    at browserslist (...\angular\aio\node_modules\browserslist\index.js:317:16)
    at Object.isSupported (...\angular\aio\node_modules\caniuse-api\dist\index.js:62:37)
    at isEs5SupportNeeded (...\angular\aio\node_modules\@angular-devkit\build-angular\src\utils\differential-loading.js:22:21)
    at Object.isDifferentialLoadingNeeded (...\angular\aio\node_modules\@angular-devkit\build-angular\src\utils\differential-loading.js:15:29)
    at generateWebpackConfig (...\angular\aio\node_modules\@angular-devkit\build-angular\src\utils\webpack-browser-config.js:25:35)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:757:11)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
```

mary-poppins commented:
You can preview a55f573 at https://pr29926-a55f573.ngbuilds.io/.

mary-poppins commented:
You can preview e62ebc8 at https://pr29926-e62ebc8.ngbuilds.io/.

gkalpak commented:
Thx to @clydin's help, I was able to overcome the error by upgrading `browserslist` to 4.5.5 (for `^4.0.0` and `^4.5.4` ranges).
So, I have upgraded @angular/cli to 8.0.0-beta.18.

For reference, here are the `browserslist` versions before the manual update to 4.5.5:
```
=> Found "browserslist@4.5.4"
info Has been hoisted to "browserslist"
info Reasons this module exists
   - Hoisted from "@angular-devkit#build-angular#caniuse-api#browserslist"
   - Hoisted from "@angular-devkit#build-angular#autoprefixer#browserslist"
=> Found "@angular-devkit/build-angular#browserslist@4.5.5"
info This module exists because "@angular-devkit#build-angular" depends on it.
```

mary-poppins commented:
You can preview 488e950 at https://pr29926-488e950.ngbuilds.io/.

AndrewKushnir commented:
Removing "cleanup" label from this PR after a chat with @gkalpak, so this PR is ready to merge.
